neqo-http3/src/connection_server.rs

```diff
@@ -235,22 +346,37 @@ impl Http3ServerHandler {
         now: Instant,
         stream_id: u64,
         buf: &mut [u8],
+        wt: bool,
     ) -> Res<(usize, bool)> {
         qinfo!([self], "read_data from stream {}.", stream_id);
-        let recv_stream = self
-            .base_handler
-            .recv_streams
-            .get_mut(&stream_id)
-            .ok_or(Error::InvalidStreamId)?
-            .http_stream()
-            .ok_or(Error::InvalidStreamId)?;
-
-        let res = recv_stream.read_data(conn, buf);
-        if let Err(e) = &res {
-            self.close(conn, now, e);
-        } else if recv_stream.done() {
-            self.base_handler.recv_streams.remove(&stream_id);
+        let (res, recv_stream) = if !wt {
+            let recv_stream = self
+                .base_handler
+                .recv_streams
+                .get_mut(&stream_id)
+                .ok_or(Error::InvalidStreamId)?;
+            let http = recv_stream.http_stream().ok_or(Error::InvalidStreamId)?;
```

Comment:
Given that we know the type of stream is involved, I think we should be conditionally extracting the stream type, as in:

```rust
            let recv_stream = self
                .base_handler
                .recv_streams
                .get_mut(&stream_id)
                .ok_or(Error::InvalidStreamId)?;
        let res = if let Some(wt) = recv_stream.wt_stream() {
           wt.read_data(conn, buf)
        } else if let Some(http) = recv_stream.http_stream() {
           http.read_data(conn, buf)
        } else {
            return Err(Error::InvalidStreamId);
        };
        match res { ...
```

Then you don't need the `wt` argument.